Use OpenDistro version of elasticsearch

alisonrclarke opened the pull request.

alisonrclarke committed:
Try running tests with opensearch instead of elasticsearch

alisonrclarke committed:
Switch to opendistro to match new cluster

alisonrclarke committed:
Add options to ES container, and add checks

alisonrclarke committed:
Check http not https

alisonrclarke committed:
Attempt to fix ES options

alisonrclarke committed:
Update docker-compose to match CI/prod

GraemeWatt reviewed (changes requested):
Looks good, thanks.  Can we run the CI on both Elasticsearch versions similar to the multiple PostgreSQL versions?  Until the `prod` cluster is also moved to Open Distro, we'll want to keep testing the code with Elasticsearch v7.1.1.

Can the [installation instructions for Elasticsearch](https://hepdata.readthedocs.io/en/latest/installation.html#id3) be modified to explain how to install Open Distro locally (not using Docker Compose), again keeping the Elasticsearch v7.1.1 instructions?  I guess Elasticsearch v7.10.2 could be installed from Homebrew, or the [Docker image from OpenDistro](https://opendistro.github.io/for-elasticsearch-docs/docs/install/docker/) could be run instead.  I got the latter to work on my laptop (adding an option `-e "opendistro_security.disabled=true"` to the `docker run` command), but when I ran `hepdata utils reindex` I got some errors like:
```
[2022-02-18 20:14:43,334: ERROR/ForkPoolWorker-3] Task hepdata.ext.elasticsearch.api.reindex_batch[e3ed776b-e23c-42bc-ad9d-fdbdb74ff721] raised unexpected: RequestError(400, 'search_phase_execution_exception', {'error': {'root_cause': [{'type': 'query_shard_exception', 'reason': '[has_parent] no join field has been configured', 'index_uuid': 'hwlsypzZTiu4ELfjXXFQMw', 'index': 'hepdata-main'}], 'type': 'search_phase_execution_exception', 'reason': 'all shards failed', 'phase': 'query', 'grouped': True, 'failed_shards': [{'shard': 0, 'index': 'hepdata-main', 'node': 'V1gOSA4IShao5hyLYOZLGw', 'reason': {'type': 'query_shard_exception', 'reason': '[has_parent] no join field has been configured', 'index_uuid': 'hwlsypzZTiu4ELfjXXFQMw', 'index': 'hepdata-main'}}]}, 'status': 400})
```
Can you please check if you get this error and see if any changes are needed when upgrading from Elasticsearch v7.1.1 to v7.10.2?  I checked that upgrading to the Python packages `elasticsearch==7.17.0` and `elasticsearch-dsl==7.4.0` gives an exception with Open Distro (`elasticsearch.exceptions.UnsupportedProductError: The client noticed that the server is not a supported distribution of Elasticsearch`) so we'll need to stick with the older versions.

alisonrclarke committed:
Add es-image to build matrix so we can check opendistro and official ES

Also removed postgres 9.6 now we've upgraded.

alisonrclarke committed:
Remove wait for ES

Shouldn't be needed for OpenDistro now we know how to set it up.
If it is needed it should only be for the OpenDistro builds.

alisonrclarke committed:
Reinstate ES check but not for plugins (which will only be on OpenDistro)

alisonrclarke committed:
Only do check for OpenDistro

alisonrclarke committed:
Attempt to run elasticsearch container manually to allow different vars

We need to pass `-e "opendistro_security.disabled=true"` to the
OpenDistro image but not the elasticsearch one. GH Actions doesn't seem
to allow that.

alisonrclarke committed:
Fix syntax error

alisonrclarke committed:
Make sure we forward ports in ES docker command

alisonrclarke committed:
Try removing ES check (again)

alisonrclarke committed:
Add debugging for ES

alisonrclarke committed:
Use an env-file for opendistro_security.disabled variable

alisonrclarke committed:
Merge branch 'master' into opensearch

alisonrclarke commented:
> ... when I ran `hepdata utils reindex` I got some errors like:
> 
> ```
> [2022-02-18 20:14:43,334: ERROR/ForkPoolWorker-3] Task hepdata.ext.elasticsearch.api.reindex_batch[e3ed776b-e23c-42bc-ad9d-fdbdb74ff721] raised unexpected: RequestError(400, 'search_phase_execution_exception', {'error': {'root_cause': [{'type': 'query_shard_exception', 'reason': '[has_parent] no join field has been configured', 'index_uuid': 'hwlsypzZTiu4ELfjXXFQMw', 'index': 'hepdata-main'}], 'type': 'search_phase_execution_exception', 'reason': 'all shards failed', 'phase': 'query', 'grouped': True, 'failed_shards': [{'shard': 0, 'index': 'hepdata-main', 'node': 'V1gOSA4IShao5hyLYOZLGw', 'reason': {'type': 'query_shard_exception', 'reason': '[has_parent] no join field has been configured', 'index_uuid': 'hwlsypzZTiu4ELfjXXFQMw', 'index': 'hepdata-main'}}]}, 'status': 400})
> ```

I think we will need to recreate the index when updating versions: `hepdata utils reindex --recreate=true` worked OK for me when switching from v7.1 to v7.10.

alisonrclarke committed:
Add installation instructions for OpenDistro/ES 7.10

alisonrclarke commented:
I think I've now got the matrix working. I took out the Postgres 9.6 now we've upgraded the production DB, so that we weren't increasing the number of runs too much. We will need to change the required checks settings when we're ready to merge.

I've also updated the installation instructions.

GraemeWatt committed:
ci: use Node version 14 to be same as Dockerfile

GraemeWatt committed:
docs: mention version 14 needed of Node.js
